### PR TITLE
Changes to support building a osquery Windows service.

### PR DESCRIPTION
**Pull request by yying**
This PR contains numerous changes to osquery to begin support for building `osqueryd.exe` as a Windows service.

### CODE REVIEW

**Comment by osqueryer** [comment]
Can one of the osquery admins reply with "ok to test" to kick off a build...

**Comment by PoppySeedPlehzr** [comment]
ok to test

**Comment by osqueryer** [comment]
Build origin/pr/2278/merge (Job: 2918) is good! :+1:

File: osquery/database/database.cpp [hunk 1]
```diff
@@ -27,7 +27,7 @@ CLI_FLAG(bool, database_dump, false, "Dump the contents of the backing store");
 
 CLI_FLAG(string,
          database_path,
-         "/var/osquery/osquery.db",
+         OSQUERY_HOME "/osquery.db",
```
Comment (theopolis) [inline osquery/database/database.cpp]: Nope! Do not put this DB into `/etc/` please.

File: osquery/extensions/extensions.cpp [hunk 2]
```diff
@@ -54,12 +54,12 @@ CLI_FLAG(bool, disable_extensions, false, "Disable extension API");
 
 CLI_FLAG(string,
          extensions_socket,
-         "/var/osquery/osquery.em",
+         OSQUERY_HOME "/osquery.em",
```
Comment (theopolis) [inline osquery/extensions/extensions.cpp]: No domain sockets in `/etc/` please!

File: osquery/logger/plugins/filesystem.cpp [hunk 3]
```diff
@@ -80,7 +80,7 @@ REGISTER(FilesystemLoggerPlugin, "logger", "filesystem");
 
 Status FilesystemLoggerPlugin::setUp() {
   log_path_ = fs::path(FLAGS_logger_path);
-
+
```
Comment (theopolis) [inline osquery/logger/plugins/filesystem.cpp]: nit, remove

File: osquery/main/windows/daemon.cpp [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+#include <osquery/core.h>
+#include <osquery/flags.h>
+#include <osquery/logger.h>
+#include <osquery/system.h>
+
+#include "osquery/core/watcher.h"
+#include "osquery/dispatcher/distributed.h"
+#include "osquery/dispatcher/scheduler.h"
+
+const std::string kServiceName = "osquery daemon service";
```
Comment (theopolis) [inline osquery/main/windows/daemon.cpp]: Could any of this go into an `osquery` namespace?

File: osquery/core/system.cpp [hunk 5]
```diff
@@ -66,6 +66,18 @@ FLAG(string,
 
 FLAG(bool, utc, false, "Convert all UNIX times to UTC");
 
+#ifdef WIN32
+struct tm *gmtime_r(time_t *t, struct tm *result) {
+	_gmtime64_s(result, t);
```
Comment (theopolis) [inline osquery/core/system.cpp]: `clang-format`

File: osquery/core/init.cpp [hunk 6]
```diff
@@ -151,14 +168,15 @@ void signalHandler(int num) {
...
+#endif
 
   if (isWatcher()) {
     // The signal should be proliferated through the process group.
     // Otherwise the watcher could 'forward' the signal to workers and
     // managed extension processes.
   }
 }
-#endif
+
```
Comment (theopolis) [inline osquery/core/init.cpp]: nit, we should end namespaces immediately following the final method.

File: osquery/core/system.cpp [hunk 7]
```diff
@@ -162,6 +174,8 @@ size_t getUnixTime() {
   return result;
 }
 
+#ifndef WIN32
+
```
Comment (theopolis) [inline osquery/core/system.cpp]: nit, nix this new line

File: osquery/main/windows/daemon.cpp [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+
+/// Logging for when we need to debug this service
+#define SLOG(...)  DebugPrintf("[osqueryd] " __VA_ARGS__)
+
+void DebugPrintf(const char *fmt, ...) {
+  va_list vl;
+  va_start(vl, fmt);
+
+  int size = _vscprintf(fmt, vl);
+  if (size) {
```
Comment (theopolis) [inline osquery/main/windows/daemon.cpp]: `if` `size` what?

File: osquery/main/windows/daemon.cpp [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+    char *buf = (char *)::HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, size + 2);
+    if (buf) {
```
Comment (theopolis) [inline osquery/main/windows/daemon.cpp]: `if` `buf` what?

File: osquery/database/database.cpp (hunk 1 above)
Comment (yying) [inline osquery/database/database.cpp]: What are your thoughts about `osquery.db` in `\ProgramData\osquery`?

Comment (theopolis) [inline osquery/database/database.cpp]: If data there can be expected to update/write often and it persists reasonably well, go to it.

**Comment by osqueryer** [comment]
Build origin/pr/2278/merge (Job: 2920) is good! :+1:

**Comment by osqueryer** [comment]
Build origin/pr/2278/merge (Job: 2923) failed :-1:

**Comment by theopolis** [comment]
:+1:

**Pull request merged**